p2p/p2p.go

```diff
@@ -264,24 +273,57 @@ func generateKey(seed int64) (crypto.PrivKey, error) {
 	return priv, nil
 }
 
-// TODO: message handling
+// handles stream; reads message length, message type, and decodes message based on type
+// TODO: implement all message types; send message back to peer when we get a message; gossip for certain message types
 func handleStream(stream net.Stream) {
 	defer func() {
 		if err := stream.Close(); err != nil {
 			log.Error("error closing stream", "err", err)
 		}
 	}()
-	// Create a buffer stream for non blocking read and write.
+
 	rw := bufio.NewReadWriter(bufio.NewReader(stream), bufio.NewWriter(stream))
-	str, err := rw.ReadString('\n')
+	lengthByte, err := rw.Reader.ReadByte()
+	if err != nil {
+		log.Error("stream handler", "got stream from", stream.Conn().RemotePeer(), "err", err)
+	}
+
+	// decode message length using LEB128
+	length := LEB128ToUint64([]byte{lengthByte})
+	log.Info("stream handler", "got message with length", length)
+
+	// read message type byte
+	msgType, err := rw.Reader.ReadByte()
 	if err != nil {
-		return
+		log.Error("stream handler", "msg type err", err)
 	}
 
-	fmt.Printf("got stream from %s: %s", stream.Conn().RemotePeer(), str)
-	_, err = rw.WriteString("hello friend")
+	// read entire message
+	msg, err := rw.Reader.Peek(int(length))
 	if err != nil {
-		return
+		log.Info("stream handler", "err", err)
+	}
+
+	log.Info("stream handler", "got stream from", stream.Conn().RemotePeer(), "message", fmt.Sprintf("%x", msg))
+
+	switch msgType {
+	case 0:
+		statusMsg := new(StatusMessage)
+		err = statusMsg.Decode(rw, length)
+		if err != nil {
+			log.Info("stream handler", "err", err)
+		}
+
+		log.Info("stream handler", "got status message from", stream.Conn().RemotePeer(),
```

Comment:
I think you can just pass the struct to the logger and get most of this info automatically, eg:
```golang
log.Info("Message", statusMsg)
```

Reply:
no it just prints out the raw struct, and also I want some of the fields to be hex formatted

Reply:
Maybe you should implement `String()` for `StatusMessage`?

Reply:
good idea I'll do that

Reply:
changes have been made